detector.go

```diff
@@ -155,92 +264,184 @@ func (c *DetectConfig) process(done []Buildpack) ([]Buildpack, []BuildPlanEntry,
 			detected = detected && bp.Optional
 		}
 	}
+
+	if len(results) > 0 {
+		detected = detected && anyUserBuildpacks(results)
+	}
+
 	if !detected {
 		if buildpackErr {
-			return nil, nil, errBuildpack
+			return nil, errBuildpack
 		}
-		return nil, nil, errFailedDetection
+		return nil, errFailedDetection
 	}
 
 	i := 0
-	deps, trial, err := results.runTrials(func(trial detectTrial) (depMap, detectTrial, error) {
+	result, err := results.runTrials(func(trial detectTrial) (*trialResult, error) {
 		i++
 		return c.runTrial(i, trial)
 	})
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
-	if len(done) != len(trial) {
-		c.Logger.Infof("%d of %d buildpacks participating", len(trial), len(done))
+	participatingBuildpacks := append(result.PrivOptions, result.BuildOptions...)
+
+	if len(done) != len(participatingBuildpacks) {
+		c.Logger.Infof("%d of %d buildpacks participating", len(participatingBuildpacks), len(done))
 	}
 
 	maxLength := 0
-	for _, t := range trial {
-		l := len(t.ID)
+	for _, bp := range participatingBuildpacks {
+		l := len(bp.ID)
 		if l > maxLength {
 			maxLength = l
 		}
 	}
 
 	f := fmt.Sprintf("%%-%ds %%s", maxLength)
 
-	for _, t := range trial {
+	for _, t := range participatingBuildpacks {
 		c.Logger.Infof(f, t.ID, t.Version)
 	}
 
-	var found []Buildpack
-	for _, r := range trial {
-		found = append(found, r.Buildpack.noOpt())
+	return &DetectResult{
+		BuildGroup:           BuildpackGroup{noOpt(result.BuildOptions)},
+		BuildPrivilegedGroup: BuildpackGroup{noOpt(result.PrivOptions)},
+		BuildPlan:            BuildPlan{noOptPlan(result.BuildDeps)},
+		RunGroup:             BuildpackGroup{noOpt(result.RunOptions)},
+		RunPlan:              BuildPlan{noOptPlan(result.RunDeps)},
+	}, nil
+}
+
+func anyUserBuildpacks(results []detectResult) bool {
+	for _, res := range results {
+		if !res.Buildpack.Privileged {
+			return true
+		}
+	}
+	return false
+}
+
+func noOpt(opts []detectOption) []Buildpack {
+	result := []Buildpack{}
+	for _, o := range opts {
+		result = append(result, o.Buildpack.noOpt())
 	}
-	var plan []BuildPlanEntry
-	for _, dep := range deps {
-		plan = append(plan, dep.BuildPlanEntry.noOpt())
+	return result
+}
+
+func noOptPlan(depMap depMap) []BuildPlanEntry {
+	var result []BuildPlanEntry
+	for _, dep := range depMap {
+		if len(dep.Providers) == 0 && len(dep.Requires) == 0 {
+			continue
+		}
+		result = append(result, dep.BuildPlanEntry.noOpt())
 	}
-	return found, plan, nil
+	return result
 }
 
-func (c *DetectConfig) runTrial(i int, trial detectTrial) (depMap, detectTrial, error) {
+type trialResult struct {
+	BuildDeps    depMap
+	BuildOptions []detectOption
+	PrivOptions  []detectOption
+	RunDeps      depMap
+	RunOptions   []detectOption
+}
+
+func (c *DetectConfig) runTrial(i int, trial detectTrial) (*trialResult, error) {
 	c.Logger.Debugf("Resolving plan... (try #%d)", i)
 
-	var deps depMap
+	buildTrial := append([]detectOption{}, trial...)
```

Comment:
What's the purpose of this line?

Reply:
Creating a copy of the `trial`. One for `build:` and one for `run:`. The methods following this modify the `trial` that is passed in and I want to ensure the same `trial` can satisfy both phases.